vpr/src/route/router_lookahead_cost_map.cpp

```diff
@@ -0,0 +1,408 @@
+    return abs(b.x() - a.x()) + abs(b.y() - a.y());
+}
+
+template<class T>
+static constexpr const T& clamp(const T& v, const T& lo, const T& hi) {
+    return std::min(std::max(v, lo), hi);
+}
+
+template<typename T>
+static vtr::Point<T> closest_point_in_rect(const vtr::Rect<T>& r, const vtr::Point<T>& p) {
+    if (r.empty()) {
+        return vtr::Point<T>(0, 0);
+    } else {
+        return vtr::Point<T>(clamp<T>(p.x(), r.xmin(), r.xmax() - 1),
+                             clamp<T>(p.y(), r.ymin(), r.ymax() - 1));
+    }
+}
+
+// build the segment map
+void CostMap::build_segment_map() {
+    const auto& device_ctx = g_vpr_ctx.device();
+    segment_map_.resize(device_ctx.rr_nodes.size());
+    for (size_t i = 0; i < segment_map_.size(); ++i) {
+        auto& from_node = device_ctx.rr_nodes[i];
+
+        int from_cost_index = from_node.cost_index();
+        int from_seg_index = device_ctx.rr_indexed_data[from_cost_index].seg_index;
+
+        segment_map_[i] = from_seg_index;
+    }
+}
+
+// resize internal data structures
+void CostMap::set_counts(size_t seg_count) {
+    cost_map_.clear();
+    offset_.clear();
+    penalty_.clear();
+    cost_map_.resize({1, seg_count});
+    offset_.resize({1, seg_count});
+    penalty_.resize({1, seg_count});
+    seg_count_ = seg_count;
+}
+
+// cached node -> segment map
+int CostMap::node_to_segment(int from_node_ind) const {
+    return segment_map_[from_node_ind];
+}
+
+static util::Cost_Entry penalize(const util::Cost_Entry& entry, int distance, float penalty) {
+    penalty = std::max(penalty, PENALTY_MIN);
+    return util::Cost_Entry(entry.delay + distance * penalty * PENALTY_FACTOR,
+                            entry.congestion, entry.fill);
+}
+
+// get a cost entry for a segment type, chan type, and offset
+util::Cost_Entry CostMap::find_cost(int from_seg_index, int delta_x, int delta_y) const {
+    VTR_ASSERT(from_seg_index >= 0 && from_seg_index < (ssize_t)offset_.size());
+    const auto& cost_map = cost_map_[0][from_seg_index];
+    if (cost_map.dim_size(0) == 0 || cost_map.dim_size(1) == 0) {
+        return util::Cost_Entry();
+    }
+
+    vtr::Point<int> coord(delta_x - offset_[0][from_seg_index].first,
+                          delta_y - offset_[0][from_seg_index].second);
+    vtr::Rect<int> bounds(0, 0, cost_map.dim_size(0), cost_map.dim_size(1));
+    auto closest = closest_point_in_rect(bounds, coord);
+    auto cost = cost_map_[0][from_seg_index][closest.x()][closest.y()];
+    float penalty = penalty_[0][from_seg_index];
+    auto distance = manhattan_distance(closest, coord);
+    return penalize(cost, distance, penalty);
+}
+
+// set the cost map for a segment type and chan type, filling holes
+void CostMap::set_cost_map(const util::RoutingCosts& delay_costs, const util::RoutingCosts& base_costs) {
+    // calculate the bounding boxes
+    vtr::Matrix<vtr::Rect<int>> bounds({1, seg_count_});
```

Comment:
Originally, this dimension was used to differentiate between CHANX and CHANY nodes, to have different cost maps for different CHAN directions. Thanks to a bug in the code, I found out that having them separated was getting worse results, both for SymbiFlow and for VTR reg-tests, therefore I kept the two CHANX/CHANY directions collapsed in one dimension only.

The CHANX and CHANY distinction comes form the normal `lookahead map` code, where the map is indeed indexed as follows:

```
map[chan_type][seg_index][delta_x][delta_y]
```

Currently, for the extended map, this dimension is virtually unused, but, to keep it as close as possible to the noraml `lookahead map` I have decided to leave it.

I think it would be better to just add a comment on why this dimension is unused.